### DISCUSSION TITLE
Keystation MK3 - Help with patch changes and control keys

### DISCUSSION THREAD

**Discussion by sedace** [GeekFunkLabs/fluidpatcher#71] [Q&A]
I seem to have setup everything according to video/web instructions with a Raspberry Pi in headless mode such that I can play piano sounds with my M-Audio Keystation MK3 controller keyboard.   Great news and thank you for all the work / efforts to this project!

One problem I have is making patch/bank changes.  I can't seem to map the control keys to function the way I'd like.   I did some searching and wanted to verify the keys were generating the right ID value as in the config.  Here's the issue:  If I use aseqdump command and monitor keypresses all the keys on my keyboard, the volume slider, and pitch bend wheel, all send midi values while watching that output.  However, the directional keys, stop, play, record buttons don't send any values when I press them and (at defaults) I just get a upper register piano note when I play them.   [Similar to this post:](https://github.com/albedozero/fluidpatcher/discussions/21)

As stated, I did some research and I tried using the Advanced button + DAW key to change the mode.  The instructions on this aren't very clear so for others who may encounter, You first press the Advanced button, release then the DAW Key (F above Middle C) to cycle through the modes.  You'll see the lights above the octave keys change colors with each press of the DAW Key....  Mackie Control - Green,     HUI - Red,    MIDI - Orange.  Once you have the desired mode you then press the "Enter" key (C key 2 octaves up) to lock the mode.  At least that's how it works on my 61 Key controller.    

Back to the topic... regardless of which mode, HUI/Mackie/Midi I am not seeing any change in the aseqdump while pressing those control keys.   Other keys do show status.  But...  Further investigation I note the Advanced DAW mode does make a functional difference.  In "green" mode the control keys just make piano sound on the audio output, they show no status in aseqdump, and make no bank/patch changes.  In "Red" mode is the same as green but they don't generate any piano sounds.  

**The main Issue/Question:** However, In "orange" mode (Midi) the left/right are actually adjusting the patches **incrementally by 2** and up / down adjust bank by 1.   --- SO CLOSE--- but I'd like to make the patch changes increment by 1 if possible or assign different control keys but I can't determine their output so any tips on this last little glitch?  

---

Unrelated Question 2:   If I wait and plug my Midi controller in after the Pi has already started it seems like I have to re-register the MIDI connection.  Is there a way to configure linux to auto assign/map "subscribe" the controller if it's unplugged/plugged into MIDI after the system has started?   Low priority, I'll probably figure this out myself over time.  

Unrelated Question 3:  Any ETA on squishbox HW being back in stock and available to order?

**Comment by albedozero** [comment]
In your main question, the MIDI mode (orange) is definitely what you want. Also - you found a bug! I had the patch change triggering on button press _and_ release. This is fixed now - if you re-run the script it will grab the latest version.

On unrelated question 1: Any MIDI controllers should be hot-pluggable, as long as `midi.autoconnect` is set to 1 in _squishboxconf.yaml_. Is this not the case for you?

Unrelated Question 2: I'm waiting for a new batch of PCBs to arrive, which should take a couple more weeks.

**Comment by sedace** [comment]
First, once again thank you for the reply and taking the time to update the code.  I updated from git.io/squishbox today for headless.py type and now the control buttons do not work at all in any of the modes (Red, Green, Orange).   aseqdump -p 28:0 (for my device) shows output for all the keys / sliders but not the control buttons on any mode... so I can't validate that buttons are the correct values but since they were mostly working prior to this update I assume that those values haven't changed and it's the update.  Might not be a bug this time, might be an issue with how the device sends data but reading through the manual I can't find anything that talks to how I configure that other than the mode type (red/green/orange) discussed above.  

Some Before:

RLS_MIDI_CHANNEL = 1
CTRL_CHAN = 1
DEC_PATCH = 21          # decrement the patch number
INC_PATCH = 22          # increment the patch number
BANK_INC = 23           # load the next bank

After:

CTRL_CHAN = 1
CHAN = 1
TYPE = 'cc'             # 'cc' or 'note'
DEC_PATCH = 21          # decrement the patch number
INC_PATCH = 22          # increment the patch number
BANK_INC = 23           # load the next bank


Maybe I'll poke around at the code and try some things on my end.  I'm likely getting in over my head though if I do...

**Reply by albedozero** [reply]
It seems strange that your Keystation isn't sending MIDI messages from the control buttons for any of the modes. Can you plug it into another computer and check with some MIDI monitoring software (like a DAW or [MIDI-OX](http://www.midiox.com/))?

**Comment by sedace** [comment]
So, I am a seasoned techie but I haven't played with MIDI in many years.  I decided to try Midi Controller and DAW again this year after MANY years away.  My apologies in advance if this is the wrong place to get support for what may simply be my being inexperienced with MIDI.  

Per your suggestion I have installed MIDI-OX and from what I can see there are two devices that get registered when I connect the controller to the computer via USB, one listed as "Keystation 61MK3" and a second one that is labeled "MIDIIN2 (Keystation 61 MK3)".  In retrospect it did appear that way in Linux/Raspberry PI as well when I was using the aseqdump -l.   I assumed I only need to use the primary one but in summary, after your suggestion to test... It appears the second interface is for the controller keys...  if I only mapped the first one the piano keys and pitchbend / modulation wheels worked and when I mapped the second MIDIIN2 then the controller key messages show up!

More importantly with the second interface connected  in MIDIOX it responds to the control keys in any mode (HUI, MACKIE, MIDI).   I'm not sure how to map that to headless.py config but I can poke around in there and maybe I'll get lucky...  for others who may encounter with both interfaces connected and enabled:

When the controller DAW is set to MIDI-Orange the output in MidiOX is Blue text.  The values for the control keys register CC for UP 23, Down 24, Left 21, Right 22.   The center button just says "Control Change" no value (but the Hex values show up as B0 14 7F and B0 14 00).  .   After each of the values 21,22,23,24 it says (E-MU) next to it.  But center button, stop (B0 76 7F and B0 76 00), play (B0 77 7F and B0 77 00), record (B0 72 7F and B0 72 00) just say control change.  

When the controller DAW is set to MACKIE-Green, the output in MidiOX is green.  The values for up C7 Note on/off, down is C#7 Note on/off, left D7, and right Eb7 respectively.  Center is E7.  Stop is A6, play is Bb6 and record is B6.  

With the controller set to HUI-RED the output is back to Blue and it says control change for all the button presses.  The HEX values are mapped differently in this mode and I don't feel like typing them all out... if someone needs I will do upon request.  

Anyway, not sure if that helped?  I'll poke around in the configs and see if I can make sure the second midi device gets mapped on boot of the raspberry pi and see if that sorts it out.
